File: Problem_Sets/HackerRank/Queuew2Stacks/Queuew2Stacks.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <stack>
using namespace std;

class Queue {
    private:
        stack<int> stack1;
        stack<int> stack2;
    
    public:
        void enqueue(int element) {
            stack1.push(element);
        }
        
        void dequeue() {
            if (stack2.empty()) {
                while (!stack1.empty()) {
                    stack2.push(stack1.top());
                    stack1.pop();
                }
            }  
            stack2.pop();
        }
        
        void peek() {
            if (!stack2.empty()) {
                cout << stack2.top() << endl;
            }
            else {
                while (!stack1.empty()) {
                    stack2.push(stack1.top());
                    stack1.pop();
                }
                cout << stack2.top() << endl;
            }
        }
        
};


int main() {
    /* Enter your code here. Read input from STDIN. Print output to STDOUT */   
    int queries;
    cin >> queries;
    
    Queue queue = Queue();
    
    for (int i = 0; i < queries; i++) {
        int fn, data;
        cin >> fn;
        cin.ignore();
        
        switch(fn) {
            case 1:
                cin >> data;
                queue.enqueue(data);
                break;
            case 2:
                queue.dequeue();
                break;
            case 3:
                queue.peek();
                break;
            default:
                cout << "Invalid Function\n";
        }
    }
    
    return 0;
}

File: Problem_Sets/HackerRank/Queuew2Stacks/Queuew2Stacks.py
# Enter your code here. Read input from STDIN. Print output to STDOUT
class Queue:
    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    def enqueue(self, data):
        self.stack1.append(data)

    def dequeue(self):
        if not self.stack2:
            while self.stack1:
                self.stack2.append(self.stack1.pop())
        self.stack2.pop()

    def peek(self):
        if self.stack2:
            return self.stack2[-1]
        else:
            while self.stack1:
                self.stack2.append(self.stack1.pop())
            return self.stack2[-1]

def main():
    queries = int(input())
    queue = Queue()

    for _ in range(queries):
        inp = input()
        if " " in inp:
            queue.enqueue(int(inp.split()[1]))
        elif inp == "2":
            queue.dequeue()
        else:
            print(queue.peek())


if __name__ == "__main__":
    main()